File: src/event/Custom.php
<?php

namespace Ething\Event;

use \Ething\Resource;

class Custom extends AbstractResourceEvent {
	
	
	static public function emit($name, Resource $resource = null){
		return new Signal('Custom', array(
			'name' => $name,
			'resource' => isset($resource) ? $resource->id() : null
		));
	}
	
	
	static public function validate(array &$attributes, array $context){
		
		$attributes = array_merge(
			array( 
				'name' => null,
				'resource' => null
			),
			$attributes
		);
		
		foreach(array_keys($attributes) as $key){
			
			switch($key){
				
				case 'name':
					if(!is_string($attributes[$key]) || empty($attributes[$key]))
						throw new \Exception("field '{$key}' must be a non empty string.");
					break;
				
				case 'resource':
					try {
						static::validateResourceAttr($context['ething'], $attributes[$key]);
					} catch(\Exception $e){
						throw new \Exception("[{$key}] ".$e->getMessage(),$e->getCode(),$e);
					}
					break;
					
				default:
					throw new \Exception("field '{$key}' unknown.");
			}
			
		}
		
		return true;
	}
	
	protected function call(Signal $signal){
		return $this->name === $signal->name && parent::call($signal);
	}
	
}

File: ething/rule/event/Custom.py
# coding: utf-8
from future.utils import string_types
from .. import Event
from .. import Signal


class Custom(Event):
    
    
    @staticmethod
    def validate (attributes, context):
        
        attributes.setdefault('name', None)
        
        for key in attributes:
            
                
            if key == 'name':
                
                name = attributes[key]
                
                if not(isinstance(name, string_types) and name ):
                    raise Exception("name must be a non empty string.");
                
            else:
                raise Exception("%s: invalid" % key)
        
        return True
    
    
    def call(self, signal):
        return signal['name'] == self['name']
    
    @classmethod
    def emit(cls, name):
        return super(Custom, cls).emit({
            'name' : name
        })
        
        
